### PR TITLE
[Feature, DEV-43] now denominate cap and goal in USD cents

### PR DESCRIPTION
**Pull request by davesag**
# major changes

* `cap` and `goal` now denominated in USD cents (using cents due to integer arithmetic)
* all dependencies updated (you will need to `npm install` to build this)
* minor updated to the docs

### CODE REVIEW

File: contracts/CTKNCrowdsale.sol [hunk 1]
```diff
@@ -138,8 +177,36 @@ contract CTKNCrowdsale is CappedCrowdsale, RefundableCrowdsale, MintedCrowdsale
      *  @param _weiAmount Value in wei to be converted into tokens
      *  @return the wei amount in excess of what was needed to buy a whole number of tokens.
      */
-    function _getOverpaymentAmount(uint256 _weiAmount) internal view returns (uint256) {
+    function _getOverpaymentAmount(uint256 _weiAmount)
+        internal
```
Comment (tavakyan) [inline contracts/CTKNCrowdsale.sol:181]: @davesag are you using a certain linter? Want to make sure I use the same one.

File: contracts/CTKNCrowdsale.sol [hunk 2]
```diff
@@ -138,8 +177,36 @@ contract CTKNCrowdsale is CappedCrowdsale, RefundableCrowdsale, MintedCrowdsale
...
+        view
+        returns (uint256)
+    {
       return _weiAmount % rate;
     }
```
Comment (tavakyan) [inline contracts/CTKNCrowdsale.sol:186]: Still think we should extend SafeMath to do _weiAmount.mod(rate)

File: contracts/CTKNCrowdsale.sol [hunk 3]
```diff
@@ -138,8 +177,36 @@ contract CTKNCrowdsale is CappedCrowdsale, RefundableCrowdsale, MintedCrowdsale
...
 
+    /**
+     * Overrised parent behavior requiring purchase to respect the funding cap but in USD
+     * @param _beneficiary Token purchaser
+     * @param _weiAmount Amount of wei contributed
+     */
+    function _preValidatePurchase(address _beneficiary, uint256 _weiAmount)
```
Comment (tavakyan) [inline contracts/CTKNCrowdsale.sol:193]: Also, not sure if working with KYC we can use WhitelistedCrowdsale: https://github.com/OpenZeppelin/zeppelin-solidity/blob/master/contracts/crowdsale/validation/WhitelistedCrowdsale.sol and override that _preValidatedPurchase . That way we can have "Shoppers" be whitelisted into investors per @smoliver diagram.

File: test/interactions/CTKNCowdsale-finalizedGoalReached_test.js [hunk 4]
```diff
@@ -24,8 +27,10 @@ contract(
     // goal reached so only refund overpayment less gas costs.
     const expectedMax = amount.minus(rate.times(2))
 
-    // TODO: The goal needs to be set in USD
-    const goal = toWei(1)
+    // Set the goal low, to 1 eth
+    // which at $850 per ETH is
+    // 85000 cents.
+    const goal = new BigNumber(85000)
```
Comment (tavakyan) [inline test/interactions/CTKNCowdsale-finalizedGoalReached_test.js]: 850 * 100 might be more readable

File: test/interactions/CTKNCowdsale-setDollarRate_test.js [hunk 5]
```diff
@@ -22,23 +22,27 @@ contract('CTKNCrowdsale setDollarRate', accounts => {
...
+    it('throws if rate is 0', () =>
+      assertThrows(crowdsale.setUSDConversionRate(0)))
   })
 
   context('given good data', () => {
-    const dollarRate = toWei(0.75)
+    // let's change it so 1 ETH is US$1000 (nice default)
+    // ie 1 dollar buys 0.001 ETH
+    // 1 cent buys 0.00001 ETH
+    const usdConversionRate = toWei(0.00001)
```
Comment (tavakyan) [inline test/interactions/CTKNCowdsale-setDollarRate_test.js:33]: const centsToWei = toWei(0.00001) ?

File: test/utils/fake.js [hunk 6]
```diff
@@ -14,20 +14,23 @@ const crowdsaleData = ({ wallet, overpaymentWallet, token, ...fields }) => {
...
 
-  // usually the goal is represented in wei but we need it to be in USD.
-  const goal = toWei(50)
+  // the default cap is 3 times the goal.
+  const cap = fields.cap || goal.times(3)
 
-  // just set a default dollar rate of half an ETH
-  const dollarRate = toWei(0.5)
+  // as a default let's say
+  // current ETH rate is approx USD 850
```
Comment (tavakyan) [inline test/utils/fake.js:24]: const ethPerUsd = 1 / 850
const ethPerCent = ethPerUsd / 100
const usdConversionRate = toWei(ethPerCent)
Not sure if this^ structure would help with additional tests

File: test/utils/fake.js [hunk 7]
```diff
@@ -14,20 +14,23 @@ const crowdsaleData = ({ wallet, overpaymentWallet, token, ...fields }) => {
...
+  // so one dollar buys 0.001176470588 ETH
+  // and thus one cent buys 0.00001176470588 ETH
+  const usdConversionRate = toWei(0.00001176470588)
 
   const data = {
     openingTime,
     closingTime,
     rate,
-    dollarRate,
+    usdConversionRate,
```
Comment (tavakyan) [inline test/utils/fake.js:33]: is it essentially centsToWeiRate?

**Review by tavakyan — COMMENTED** [review]
This looks good

File: test/utils/fake.js (hunk 7 above)
Comment (davesag) [reply test/utils/fake.js:33]: yes but cents denominated in USD currency

File: test/interactions/CTKNCowdsale-setDollarRate_test.js (hunk 5 above)
Comment (davesag) [reply test/interactions/CTKNCowdsale-setDollarRate_test.js:33]: `cents` could be one of a hundred currencies

File: contracts/CTKNCrowdsale.sol (hunk 3 above)
Comment (davesag) [reply contracts/CTKNCrowdsale.sol:193]: In our case the whitelist is attached to the token and not the crowdsale (or maybe it needs to be attached to both?)

Comment (davesag) [reply contracts/CTKNCrowdsale.sol:193]: either way that can happen in a future PR

File: contracts/CTKNCrowdsale.sol (hunk 2 above)
Comment (davesag) [reply contracts/CTKNCrowdsale.sol:186]: invoking library functions uses more gas than straight up arithmetic as it uses the `delegate_call` opcode under the hood.  I don't see a reason to do that for this.

File: contracts/CTKNCrowdsale.sol (hunk 1 above)
Comment (davesag) [reply contracts/CTKNCrowdsale.sol:181]: using `solhint`

**Pull request merged**